arwen/cryptoapi/cryptoei.go

```diff
@@ -0,0 +1,266 @@
+// extern int32_t keccak256(void *context, int32_t dataOffset, int32_t length, int32_t resultOffset);
+// extern int32_t ripemd160(void *context, int32_t dataOffset, int32_t length, int32_t resultOffset);
+// extern int32_t verifyBLS(void *context, int32_t keyOffset, int32_t messageOffset, int32_t messageLength, int32_t sigOffset);
+// extern int32_t verifyEd25519(void *context, int32_t keyOffset, int32_t messageOffset, int32_t messageLength, int32_t sigOffset);
+// extern int32_t verifySecp256k1(void *context, int32_t keyOffset, int32_t keyLength, int32_t messageOffset, int32_t messageLength, int32_t sigOffset);
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/wasmer"
+)
+
+const blsPublicKeyLength = 96
+const blsSignatureLength = 48
+const ed25519PublicKeyLength = 32
+const ed25519SignatureLength = 64
+const secp256k1CompressedPublicKeyLength = 33
+const secp256k1UncompressedPublicKeyLength = 65
+const secp256k1SignatureLength = 64
+
+func CryptoImports(imports *wasmer.Imports) (*wasmer.Imports, error) {
+	imports = imports.Namespace("env")
+	imports, err := imports.Append("sha256", sha256, C.sha256)
+	if err != nil {
+		return nil, err
+	}
+
+	imports, err = imports.Append("keccak256", keccak256, C.keccak256)
+	if err != nil {
+		return nil, err
+	}
+
+	imports, err = imports.Append("ripemd160", ripemd160, C.ripemd160)
+	if err != nil {
+		return nil, err
+	}
+
+	imports, err = imports.Append("verifyBLS", verifyBLS, C.verifyBLS)
+	if err != nil {
+		return nil, err
+	}
+
+	imports, err = imports.Append("verifyEd25519", verifyEd25519, C.verifyEd25519)
+	if err != nil {
+		return nil, err
+	}
+
+	imports, err = imports.Append("verifySecp256k1", verifySecp256k1, C.verifySecp256k1)
+	if err != nil {
+		return nil, err
+	}
+
+	return imports, nil
+}
+
+//export sha256
+func sha256(context unsafe.Pointer, dataOffset int32, length int32, resultOffset int32) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.SHA256
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	result, err := crypto.Sha256(data)
+	if err != nil {
+		return 1
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	return 0
+}
+
+//export keccak256
+func keccak256(context unsafe.Pointer, dataOffset int32, length int32, resultOffset int32) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.SHA256
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
```

Comment:
Consume `DataCopyPerByte` for each byte of `data`, before loading it from memory (`data` has variable length).

Reply:
Done